In [1]:
from scipy.signal import residuez

## Άσκηση - Ανάπτυγμα σε Μερικά Κλάσματα στην Python (Μετασχηματισμός $Z$)

Είδαμε ότι το Ανάπτυγμα σε Μερικά Κλάσματα είναι μια σωτήρια 😃 τεχνική όσον αφορά αντίστροφους μετασχηματισμούς που εκφράζονται ως **ρητή** συνάρτηση. Ό,τι κάναμε για τον μετασχηματισμό Laplace με τη συνάρτηση ``residue``, μπορούμε να το κάνουμε και για τον μετασχηματισμό $Z$ με την αντίστοιχη συνάρτηση ``residuez`` της $\texttt{SciPy}$.

Η σύνταξη της εντολής για να μας κάνει το ανάπτυγμα είναι η ακόλουθη:

$$\textrm{r, p, k = residuez(b, a)}$$

**Προσοχή σε μια πολύ σημαντική διαφορά με την ``residue``!** Η ``residuez`` δουλεύει με τη ρητή συνάρτηση εκφρασμένη σε **δυνάμεις του** $z^{-1}$, δηλαδή στη μορφή

$$X(z) = \frac{b_0 + b_1 z^{-1} + b_2 z^{-2} + \cdots}{a_0 + a_1 z^{-1} + a_2 z^{-2} + \cdots}$$

Επομένως:

- $\textrm{b}$: ένα διάνυσμα με τους συντελεστές του πολυωνύμου του **αριθμητή**, σε **αύξουσα** σειρά δυνάμεων του $z^{-1}$ (πρώτα ο σταθερός όρος $z^{0}$, μετά ο συντελεστής του $z^{-1}$, μετά του $z^{-2}$, κ.ο.κ.). Προσέξτε ότι είναι το **αντίθετο** από ό,τι κάναμε στην ``residue``, όπου βάζαμε τους συντελεστές του $s$ σε *φθίνουσα* σειρά βαθμού!
- $\textrm{a}$: ένα διάνυσμα με τους συντελεστές του πολυωνύμου του **παρονομαστή**, επίσης σε **αύξουσα** σειρά δυνάμεων του $z^{-1}$.

Η συνάρτηση επιστρέφει τρία διανύσματα, με τα στοιχεία καθενός σε αντιστοιχία με των υπολοίπων:

- το διάνυσμα $\textrm{r}$ περιέχει τους αριθμητές των μερικών κλασμάτων που προκύπτουν,
- το διάνυσμα $\textrm{p}$ περιέχει τους πόλους των απλών κλασμάτων, και
- το διάνυσμα $\textrm{k}$ περιέχει τους όρους που προκύπτουν από τη διαίρεση των πολυωνύμων, στην περίπτωση που ο βαθμός του αριθμητή $M$ είναι μεγαλύτερος ή ίσος με αυτόν του παρονομαστή, $N$ (δηλ. $M \geq N$).

Μόνο που τώρα κάθε μερικό κλάσμα έχει τη μορφή

$$\frac{r_i}{1 - p_i\, z^{-1}}$$

Αυτή είναι ακριβώς η βολική μορφή της οποίας ο (**αιτιατός**) αντίστροφος μετασχηματισμός $Z$ είναι $r_i\, p_i^{\,n}\, u[n]$.

Ας δούμε δυο παραδείγματα.

1. Έστω:

$$X(z) = \frac{1 + z^{-1}}{1 - 0.75\,z^{-1} + 0.125\,z^{-2}}$$

Ο αριθμητής έχει συντελεστές πολυωνύμου τους $[1, 1]$ και ο παρονομαστής $[1, -0.75, 0.125]$ αντίστοιχα (σε αύξουσες δυνάμεις του $z^{-1}$).

Εκτελούμε λοιπόν:

In [2]:
b = [1, 1]
a = [1, -0.75, 0.125]

r, p, k = residuez(b, a)

print(f"Αριθμητές = {r} \nΠόλοι = {p} \nΌροι συναρτήσεων δέλτα = {k}")

Αριθμητές = [ 6. -5.] 
Πόλοι = [0.5  0.25] 
Όροι συναρτήσεων δέλτα = []


Επειδή το διάνυσμα $k$ είναι άδειο, δεν υπάρχουν όροι από τη διαίρεση των πολυωνύμων - λογικό κι αναμενόμενο, καθώς ο αριθμητής είναι πρώτου βαθμού (σε $z^{-1}$) και ο παρονομαστής δευτέρου, οπότε δε χρειάστηκε διαίρεση.

Από τα υπόλοιπα, παρατηρούμε ότι μπορούμε να γράψουμε το ανάπτυγμα ως

$$X(z) = \frac{1 + z^{-1}}{1 - 0.75\,z^{-1} + 0.125\,z^{-2}} = (6)\frac{1}{1 - \frac{1}{2}z^{-1}} + (-5)\frac{1}{1 - \frac{1}{4}z^{-1}}$$

Δυστυχώς, εδώ η ``Python`` μας αφήνει. 😃 Στη συνέχεια, ανάλογα με το πεδίο σύγκλισης (ROC), γυρνάμε τα σήματά μας πίσω στο χρόνο μέσω των γνωστών πινάκων μετασχηματισμού $Z$. Για **αιτιατό** σήμα (δεξιόπλευρο πεδίο σύγκλισης, $|z| > \frac{1}{2}$):

$$x[n] = 6\left(\tfrac{1}{2}\right)^{n} u[n] - 5\left(\tfrac{1}{4}\right)^{n} u[n]$$

2. Έστω:

$$X(z) = \frac{1 + 2z^{-1} + 3z^{-2} + z^{-3}}{1 - 0.75\,z^{-1} + 0.125\,z^{-2}}$$

Ο αριθμητής έχει συντελεστές $[1, 2, 3, 1]$ και ο παρονομαστής $[1, -0.75, 0.125]$ αντίστοιχα. Παρατηρούμε ότι ο αριθμητής (3ου βαθμού σε $z^{-1}$) έχει μεγαλύτερο βαθμό από τον παρονομαστή (2ου), δηλ. $M > N$, οπότε θα προκύψουν όροι από τη διαίρεση των πολυωνύμων. Αυτό η ``Python`` το κάνει για μας!

Εκτελούμε λοιπόν:

In [3]:
b = [1, 2, 3, 1]
a = [1, -0.75, 0.125]

r, p, k = residuez(b, a)

print(f"Αριθμητές = {r} \nΠόλοι = {p} \nΌροι συναρτήσεων δέλτα = {k}")

Αριθμητές = [  50. -121.] 
Πόλοι = [0.5  0.25] 
Όροι συναρτήσεων δέλτα = [72.  8.]


Τώρα το διάνυσμα $k = [72,\ 8]$ ΔΕΝ είναι άδειο. Παρατηρούμε ότι μπορούμε να γράψουμε το ανάπτυγμα ως

$$X(z) = \frac{1 + 2z^{-1} + 3z^{-2} + z^{-3}}{1 - 0.75\,z^{-1} + 0.125\,z^{-2}} = 72 + 8z^{-1} + (50)\frac{1}{1 - \frac{1}{2}z^{-1}} + (-121)\frac{1}{1 - \frac{1}{4}z^{-1}}$$

Γιατί ονόμασα το διάνυσμα $[72,\ 8]$ ως "όρους συναρτήσεων δέλτα"; Οι όροι $72 + 8z^{-1}$ είναι ένα πολυώνυμο του $z^{-1}$, και ο αντίστροφος μετασχηματισμός $Z$ του $z^{-m}$ είναι η $\delta[n-m]$. Άρα, για αιτιατό σήμα:

$$x[n] = 72\,\delta[n] + 8\,\delta[n-1] + 50\left(\tfrac{1}{2}\right)^{n} u[n] - 121\left(\tfrac{1}{4}\right)^{n} u[n]$$

**Παρατηρήστε τη διαφορά με τον Laplace:** εκεί οι όροι του $k$ έδιναν τη $\delta(t)$ και τις *παραγώγους* της (π.χ. $\frac{d}{dt}\delta(t)$)· εδώ, στο πεδίο του $Z$, οι όροι του $k$ δίνουν απλώς *χρονικά ολισθημένες* κρουστικές $\delta[n], \delta[n-1], \dots$ - πολύ πιο βατό! Όπως και στον Laplace, δε συνηθίζεται να έχουμε συστήματα με $M \geq N$, οπότε εδώ απλά χρησιμοποιήσαμε ένα τέτοιο για λόγους πληρότητας της επίδειξης της ``residuez``.

---

Αναπτύξτε σε μερικά κλάσματα τους παρακάτω μετασχηματισμούς $Z$ (θεωρήστε αιτιατά σήματα):

(α') $$X(z) = \frac{1}{(1 - z^{-1})(1 - 0.5\,z^{-1})}$$
(β') $$X(z) = \frac{1}{1 - 0.25\,z^{-2}}$$
(γ') $$X(z) = \frac{1}{(1 - 0.5\,z^{-1})^{2}}$$
(δ') $$X(z) = \frac{1 + z^{-3}}{1 - 0.25\,z^{-2}}$$

### Απαντήσεις:

In [4]:
# (α')
b = [1]
a = [1, -1.5, 0.5]

r, p, k = residuez(b, a)

print(f"Αριθμητές = {r} \nΠόλοι = {p} \nΌροι συναρτήσεων δέλτα = {k}")

# Πρέπει να βρείτε
# r = [ 2. -1.] , p = [1.  0.5] , k = []
# δηλ.  x[n] = 2 u[n] - (1/2)^n u[n]


Αριθμητές = [ 2. -1.] 
Πόλοι = [1.  0.5] 
Όροι συναρτήσεων δέλτα = []


In [5]:
# (β')
b = [1]
a = [1, 0, -0.25]

r, p, k = residuez(b, a)

print(f"Αριθμητές = {r} \nΠόλοι = {p} \nΌροι συναρτήσεων δέλτα = {k}")

# Πρέπει να βρείτε
# r = [0.5 0.5] , p = [ 0.5 -0.5] , k = []
# δηλ.  x[n] = (1/2)[ (1/2)^n + (-1/2)^n ] u[n]


Αριθμητές = [0.5 0.5] 
Πόλοι = [ 0.5 -0.5] 
Όροι συναρτήσεων δέλτα = []


In [6]:
# (γ')
b = [1]
a = [1, -1, 0.25]

r, p, k = residuez(b, a)

print(f"Αριθμητές = {r} \nΠόλοι = {p} \nΌροι συναρτήσεων δέλτα = {k}")

# Πρέπει να βρείτε
# r = [0. 1.] , p = [0.5 0.5] , k = []
# Διπλός πόλος: ο δεύτερος όρος αντιστοιχεί στο 1/(1 - 0.5 z^-1)^2
# δηλ.  x[n] = (n+1)(1/2)^n u[n]


Αριθμητές = [0. 1.] 
Πόλοι = [0.5 0.5] 
Όροι συναρτήσεων δέλτα = []


In [7]:
# (δ')
b = [1, 0, 0, 1]
a = [1, 0, -0.25]

r, p, k = residuez(b, a)

print(f"Αριθμητές = {r} \nΠόλοι = {p} \nΌροι συναρτήσεων δέλτα = {k}")

# Πρέπει να βρείτε
# r = [ 4.5 -3.5] , p = [ 0.5 -0.5] , k = [ 0. -4.]
# δηλ.  x[n] = -4 δ[n-1] + 4.5(1/2)^n u[n] - 3.5(-1/2)^n u[n]


Αριθμητές = [ 4.5 -3.5] 
Πόλοι = [ 0.5 -0.5] 
Όροι συναρτήσεων δέλτα = [ 0. -4.]


---
---